R/stringify_expression.R

```diff
@@ -1,66 +1,60 @@
+
+  if (is.null(.srcref)) {
+    return(deparse(x))
   }
 
-  ## https://journal.r-project.org/archive/2010-2/RJournal_2010-2_Murdoch.pdf
-  ## Each element is a vector of 6 integers: (first line, first byte, last line,
-  ## last byte, first character, last character)
-  ## ^^ clearly not whole story because I see vectors of length 8
-  ## but seems to be true description of first 6 elements
-  coord <- do.call(rbind, lapply(src, `[`, c(1, 5, 3, 6)))
-  colnames(coord) <- c("first_line", "first_character",
-                       "last_line", "last_character")
-  n_statements <- nrow(coord)
-  ## each row holds coordinates of source for one statement
+  ## Construct a new srcref with the first_line, first_byte, etc. from the
+  ## first expression and the last_line, last_byte, etc. from the last one.
+  first_src <- .srcref[[1]]
+  last_src <- .srcref[[length(.srcref)]]
 
-  ## all statements seem to contain such a reference to same lines of source
-  lines <-  attr(src[[1]], "srcfile")$lines
+  .srcfile <- attr(first_src, "srcfile")
 
-  ## make sure source is split into character vector anticipated by coord
-  if (!isTRUE(attr(src[[1]], "srcfile")$fixedNewlines)) {
-    lines <- strsplit(lines, split = "\n")[[1]]
-  }
+  src <- srcref(
+    .srcfile,
+    c(first_src[[1]], first_src[[2]],
+       last_src[[3]],  last_src[[4]],
+      first_src[[5]],  last_src[[6]],
+      first_src[[7]],  last_src[[8]])
+  )
 
-  lines <- lines[coord[1, "first_line"]:coord[n_statements, "last_line"]]
+  lines <- as.character(src, useSource = TRUE)
 
-  ## line 1 needs to be pruned, possibly from both ends
-  ## default is this in order to retain end of line comments
-  line_1_end <- nchar(lines[1])
-  ## but still have to worry about multiple statements in one line, eg {1:5;6:9}
-  if (n_statements > 1 && coord[2, "last_line", drop = TRUE] == 1) {
-    line_1 <- coord[ , "last_line", drop = TRUE] == 1
-    line_1_end <- max(coord[line_1, "last_character"])
+  ## remove the first brace and line if the brace is the only thing on the line
+  lines <- sub("^[{]", "", lines)
+  if (!nzchar(lines[[1L]])) {
+    lines <- lines[-1L]
   }
-  lines[1] <- substr(lines[1], coord[1, "first_character"], line_1_end)
-  lines[1] <- sub("^\\{", "", lines[1])
 
-  ## last line may also need to be truncated
-  n_lines <- length(lines)
-  ## but don't do anything if line 1 == last line
-  if (n_statements > 1 && n_lines > 1) {
-    lines[n_lines] <- substr(lines[n_lines],
-                             coord[n_statements, "first_character"],
-                             coord[n_statements, "last_character"])
-  }
+  ## rescue trailing comment lines
+  ## n = the last line consulted thus far
+  ## FYI the 3rd element corresponds to last_line
+  n <- max(vapply(.srcref, function(x) x[[3]], integer(1)))
+  tail_lines <- trailing_comments(.srcfile, n)
+
+  c(lines, tail_lines)
+}
+
+trailing_comments <- function(.srcfile, n) {
+  raw_lines <- .srcfile[["lines"]]
```

Comment:
I am pretty sure it would be better to use `utils::getSrcLines(.srcfile, 1, Inf)` to retrieve the lines.

Reply:
Indeed!